### PR TITLE
Deepseek v4 csa mask collapse

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?

See https://github.com/huggingface/transformers/pull/45892#issuecomment-4432108446 I think its worth having

### CODE REVIEW

File: src/transformers/models/deepseek_v4/modeling_deepseek_v4.py [hunk 1]
```diff
@@ -679,26 +679,15 @@ def forward(
...
         # Per-query block bias: query `t` may only see the cache entries that are <= `seq_len // m`
         # and in these, only the ones marked valid by the indexer. Everything else is `-inf`.
         # While the above negated the indexer, here we apply the "causal" masking.
-        block_bias = gathered.new_full((batch, 1, seq_len, seq_len, top_k), float("-inf"))
-        allowed = torch.where(valid, gathered.new_zeros(()), gathered.new_full((), float("-inf")))  # [B, S, k]
-        query_indices = torch.arange(seq_len, device=gathered.device)
-        block_bias[:, 0, query_indices, query_indices, :] = allowed  # diagonal: q_idx == block_idx
-        block_bias = block_bias.view(batch, 1, seq_len, seq_len * top_k)
-        return gathered, block_bias
+        safe_indices = torch.where(valid, top_k_indices, torch.full_like(top_k_indices, compressed_len))
```
Comment (vadimkantorov) [inline src/transformers/models/deepseek_v4/modeling_deepseek_v4.py:687]: Is `full_like` needed? Modern `torch.where` can handle scalar arguments / or broadcast directly

Comment (vasqu) [reply src/transformers/models/deepseek_v4/modeling_deepseek_v4.py:687]: my immediate thought is that it can cause device syncs if you use normal scalars so it's likely easier to create the tensor at that scale
